src/plugin.c

```diff
@@ -86,9 +87,11 @@ int plugin_register(plugin_t *plugin)
 	if (!plugin->name) {
 #ifndef ENABLE_STATIC
 		Dl_info info;
+		char *dli_fname = strdup(info.dli_fname);
 
-		if (dladdr(plugin, &info) && info.dli_fname)
-			plugin->name = basename(info.dli_fname);
+		if (dladdr(plugin, &info) && dli_fname)
+			plugin->name = basename(dli_fname);
+		free(dli_fname);
```

Comment:
Why this code?  There is no mention of it in the PR, the commit message, or the original issue filing explaining why you believe it's needed.

### Code review

 1. You do not check the return value from `strdup()`, it may be `NULL`
 2. You assign `plugin->name` a subset of the allocated memory and then `free()` it?  That means `plugin->name` will point to a region of memory that may be reused soon for other things.
 3. You don't need to `strdup()`, the `basename()` function finds the last `/` (if any) and returns a pointer the the char right after it.  No other use of `basename()` in the code uses a `strdup()`.

### Final comment

Also, you missed one use of `basename()` in `plugin/modprobe.c`.  Not sure if one of the other includes in turn include `libgen.h`, or if your build does not have the modprobe plugin enabled.

Reply:
Oh apologies for forgetting to mention it anywhere.
Why is that here:
- posix basename expects to be given a `char*` instead of a `const char*`, in fact the [spec](https://pubs.opengroup.org/onlinepubs/007904975/functions/basename.html) and the man page for [basename(3)](https://man7.org/linux/man-pages/man3/basename.3.html) mention, that the given string may be modified. This would result in undefined behavior if it tries to modify a read-only memory address.
> The basename() function may modify the string pointed to by path, and may return a pointer to static storage that may then be overwritten by a subsequent call to basename().

> Both dirname() and basename() may modify the contents of path, so
       it may be desirable to pass a copy when calling one of these
       functions.
       
- The goal was to create a copy of `info.dli_fname` that is allowed to be modified to adhere to the spec

### Review answers
1. Good catch, it went over my head that `strdup()` may return `NULL`
2. Yes the `free()` makes absolutely no sense there
3. Others uses of `basename()` in the code base do not require the usage of `strdup()`, because they are all a char pointer, opposed to this specific instance, where it is a `const char*`

### Final comment answers
`plugin/modprobe.c` does not have a call to `basename()`? It has a local variable named `basename` in the `maybe_fopen_alias()` function.
But I forgot one in `tmpfiles.c`

Reply:
OK, this is a mess. I've just had a look at the source implementation of `basename()` in both musl and glibc.  The xpg versions do indeed modify the contents, even though there's no need for it.  The GNU version of `basename()` in GLIBC, however, does not. So I guess that's what I looked the last time I dug into this.

I think for consistency, and to keep things sane, we need to carry our own version of `basename()`, e.g., `basenm()` that basically does what parts of `maybe_fopen_alias()` does.  That would be my solution.

Reply:
I think that is the best solution going forth. Would a suitable place for the custom `basenm()` function be `src/helpers.c`?

Reply:
Put it with `progname()` and `str()` in `src/util.c`, we may very possibly want to reuse this in `initctl` and other programs later.  Thanks!